src/metabase/api/dashboard.clj

```diff
@@ -607,6 +564,92 @@
+         (api/maybe-reconcile-collection-position! dash-before-update dash-updates)
+         (when-let [updates (not-empty
+                             (u/select-keys-when
+                              dash-updates
+                              :present #{:description :position :collection_id :collection_position :cache_ttl}
+                              :non-nil #{:name :parameters :caveats :points_of_interest :show_in_getting_started :enable_embedding
+                                         :embedding_params :archived :auto_apply_filters}))]
+           (t2/update! Dashboard id updates))
+         (when update-dashcards-and-tabs?
+           (let [new-tabs                 (map-indexed (fn [idx tab] (assoc tab :position idx)) tabs)
+                 {:keys [old->new-tab-id
+                         deleted-tab-ids]
+                  :as tabs-changes-stats} (dashboard-tab/do-update-tabs! (:id dash-before-update) (:tabs dash-before-update) new-tabs)
+                 deleted-tab-ids          (set deleted-tab-ids)
+                 current-cards            (cond->> (:dashcards dash-before-update)
+                                            (seq deleted-tab-ids)
+                                            (remove (fn [dashcard]
+                                                      (contains? deleted-tab-ids (:dashboard_tab_id dashcard)))))
+                 new-cards                (cond->> dashcards
+                                            ;; fixup the temporary tab ids with the real ones
+                                            (seq old->new-tab-id)
+                                            (map (fn [card]
+                                                   (if-let [real-tab-id (get old->new-tab-id (:dashboard_tab_id card))]
+                                                     (assoc card :dashboard_tab_id real-tab-id)
+                                                     card))))
+                 dashcards-changes-stats  (do-update-dashcards! dash-before-update current-cards new-cards)]
+             (reset! changes-stats
+                     (merge
+                      (select-keys tabs-changes-stats [:created-tab-ids :updated-tab-ids :deleted-tab-ids :total-num-tabs])
+                      (select-keys dashcards-changes-stats [:created-dashcards :deleted-dashcards :updated-dashcards]))))))
+       true))
+    (let [dashboard (t2/select-one :model/Dashboard id)]
+      (when update-dashboard-itself?
+        ;; execute these events for old PUT /api/dashboard/:id calls and new PUT /api/dashboard/:id calls, and not for old PUT/api/dashboard/:id/card calls
+        (events/publish-event! :event/dashboard-update {:object dashboard :user-id api/*current-user-id*}))
+      (when update-dashcards-and-tabs?
+        ;; execute these events for old PUT /api/dashboard/:id/cards calls and new PUT /api/dashboard/:id calls, and not for old PUT /api/dashboard/:id calls
```

Comment:
Here I'm being careful not to make any breaking changes. 

That means:
1. If the FE makes a request to `PUT /api/dashboard/:id` with no dashcards or tabs in the body, followed by `PUT /api/dashboard/:id/cards` with dashcards and tabs in the body, the same set of events should be fired before and after this PR.
AND
2. If the FE makes a single request to `PUT /api/dashboard/:id` with dashcards or tabs in the body, the same set of events should be fired as in 1.

To understand these conditions are met, you need to know:
```
(track-dashcard-and-tab-events! ...) used to be in the PUT /api/dashboard/:id/cards handler.
(events/publish-event! ...) used to be in the PUT /api/dashboard/:id handler.
```
and then read the code